pkg/azuredisk/nodeserver_v2.go

```diff
@@ -90,9 +97,52 @@ func (d *DriverV2) NodeStageVolume(ctx context.Context, req *csi.NodeStageVolume
 		return nil, status.Errorf(codes.Internal, "Failed to find disk on lun %s. %v", lunStr, err)
 	}
 
-	source, err := d.nodeProvisioner.GetDevicePathWithLUN(int(lun))
+	// check if AzVolumeattachment is in Attached state if the disk detachment is not in process
+	if state, getErr := d.crdProvisioner.GetAzVolumeAttachmentState(ctx, diskURI, d.NodeID); getErr != nil {
+		getErr = status.Errorf(codes.Internal, "failed to get current attachment state for volume (%s) and node (%s): %v", diskURI, d.NodeID, getErr)
+		klog.Error(getErr)
+		return nil, getErr
+	} else if !d.nodeProvisioner.IsDetachInProcess(diskURI) {
+		getErr = status.Errorf(codes.Internal, "recovery for volume (%s) is still in process", diskURI)
+		klog.Error(getErr)
+		return nil, getErr
+	} else if state != v1alpha1.Attached {
+		getErr = status.Errorf(codes.Internal, "volume (%s) is not yet attached to node (%s)", diskURI, d.NodeID)
+		klog.Error(getErr)
+		return nil, getErr
+	}
```

Comment:
We should only need to check the `AzVolumeAttachment` state if a detach/recovery has been started. Once we see the state is `Attached` then we can clear the "detach in progress" state and proceed. It's OK to use the informer cached state,  since `CrdProvisioner.UnpublishVolume` will have waited until the informer cache has updated the `AzVolumeAttachment` state to `Detached`.